pyro/poutine/runtime.py

```diff
@@ -339,18 +338,18 @@ def effectful(
 
     @functools.wraps(fn)
     def _fn(
-        *args: P.args,
+        *args: _P.args,
         name: Optional[str] = None,
         infer: Optional[InferDict] = None,
-        obs: Optional[torch.Tensor] = None,
-        **kwargs: P.kwargs,
-    ) -> Union[T, torch.Tensor, None]:
+        obs: Optional[_T] = None,
+        **kwargs: _P.kwargs,
+    ) -> Optional[_T]:
```

Comment:
Changed this back to return `Optional` and removed the `assert  msg["value"] is not None` line. One concern I have is that if `_T` itself is `None` then it will raise an assertion error.